src/java_bytecode/ci_lazy_methods.cpp

```diff
@@ -122,46 +122,88 @@ bool ci_lazy_methodst::operator()(
 
-  bool any_new_methods = true;
-  while(any_new_methods)
+  bool any_new_classes = true;
+  while(any_new_classes)
   {
-    any_new_methods=false;
-    while(!methods_to_convert_later.empty())
+    bool any_new_methods = true;
+    while(any_new_methods)
     {
-      id_sett methods_to_convert;
-      std::swap(methods_to_convert, methods_to_convert_later);
-      for(const auto &mname : methods_to_convert)
+      any_new_methods = false;
+      while(!methods_to_convert_later.empty())
       {
-        if(!methods_already_populated.insert(mname).second)
-          continue;
-        debug() << "CI lazy methods: elaborate " << mname << eom;
-        if(
-          method_converter(
-            mname,
-            // Note this wraps *references* to methods_to_convert_later &
-            // instantiated_classes
-            ci_lazy_methods_neededt(
-              methods_to_convert_later, instantiated_classes, symbol_table)))
+        id_sett methods_to_convert;
+        std::swap(methods_to_convert, methods_to_convert_later);
+        for(const auto &mname : methods_to_convert)
         {
-          // Couldn't convert this function
-          continue;
+          if(!methods_already_populated.insert(mname).second)
+            continue;
+          debug() << "CI lazy methods: elaborate " << mname << eom;
+          if(
+            method_converter(
+              mname,
+              // Note this wraps *references* to methods_to_convert_later &
+              // instantiated_classes
+              ci_lazy_methods_neededt(
+                methods_to_convert_later, instantiated_classes, symbol_table)))
+          {
+            // Couldn't convert this function
+            continue;
+          }
+          gather_virtual_callsites(
+            symbol_table.lookup_ref(mname).value, virtual_function_calls);
+          any_new_methods = true;
         }
-        gather_virtual_callsites(
-          symbol_table.lookup_ref(mname).value, virtual_function_calls);
-        any_new_methods=true;
       }
-    }
 
-    // Given the object types we now know may be created, populate more
-    // possible virtual function call targets:
+      // Given the object types we now know may be created, populate more
+      // possible virtual function call targets:
 
-    debug() << "CI lazy methods: add virtual method targets ("
-            << virtual_function_calls.size() << " callsites)" << eom;
+      debug() << "CI lazy methods: add virtual method targets ("
+              << virtual_function_calls.size() << " callsites)" << eom;
 
-    for(const exprt &function : virtual_function_calls)
+      for(const exprt &function : virtual_function_calls)
+      {
+        get_virtual_method_targets(
+          function,
+          instantiated_classes,
+          methods_to_convert_later,
+          symbol_table);
+      }
+    }
+
+    any_new_classes = false;
+
+    // Find virtual callsites with no candidate targets, guess that the class
```

Comment:
Suggest giving example: if we have `invokevirtual A.f` and we don't believe either `A` or any of its children may exist, assume specifically `A` is somehow instantiated. Note this may result in an abstract class being classified as instantiated, which stands in for some unknown concrete subclass: in this case the called method will be a stub.